Plutarch/Builtin.hs

```diff
@@ -68,9 +69,46 @@ pheadBuiltin = phoistAcyclic $ pforce $ punsafeBuiltin PLC.HeadList
 ptailBuiltin :: Term s (PBuiltinList a :--> PBuiltinList a)
 ptailBuiltin = phoistAcyclic $ pforce $ punsafeBuiltin PLC.TailList
 
+pchooseListBuiltin :: Term s (PBuiltinList a :--> b :--> b :--> b)
+pchooseListBuiltin = pforce $ pforce $ punsafeBuiltin PLC.ChooseList
+
 pnullBuiltin :: Term s (PBuiltinList a :--> PBool)
 pnullBuiltin = phoistAcyclic $ pforce $ punsafeBuiltin PLC.NullList
 
+pconsBuiltin :: Term s (a :--> PBuiltinList a :--> PBuiltinList a)
+pconsBuiltin = pforce $ punsafeBuiltin PLC.MkCons
+
+--------------------------------------------------------------------------------
+
+instance PLC.DefaultUni `PLC.Contains` PHaskellType a => PlutusType (PBuiltinList a) where
+  type PInner (PBuiltinList a) b = PBuiltinList a
+  pcon' :: forall s. PBuiltinList a s -> forall b. Term s (PInner (PBuiltinList a) b)
+  pcon' (PCons x xs) = pconsBuiltin # x # (pto xs)
+  pcon' PNil = pconstant @(PBuiltinList a) []
+  pmatch' xs f =
+    pforce $
+      pchooseListBuiltin
+        # xs
+        # pdelay (f PNil)
+        # pdelay (f (PCons (pheadBuiltin # xs) (punsafeFrom $ ptailBuiltin # xs)))
+
+class PLC.Contains PLC.DefaultUni (PHaskellType a) => InDefaultUni a
+instance PLC.Contains PLC.DefaultUni (PHaskellType a) => InDefaultUni a
+
+instance PListLike PBuiltinList where
+  type PElemConstraint PBuiltinList = InDefaultUni
+  pelimList match_cons match_nil =
+    plam $ \ls -> pmatch ls $ \case
+      PCons x xs -> match_cons # x # xs
+      PNil -> pforce match_nil
+  pconsList = plam $ \x xs -> pcon (PCons x xs)
+  pnilList = pcon PNil
+  punsafeHead = pheadBuiltin
```

Comment:
Note: These aren't unsafe. Unsafety refers to unsoundness, when you have a type for a term that is not within that type, e.g. `punsafeCoerce (pcon PTrue) :: ClosedTerm PInteger`.